Include transform in each unit export

rkusa opened the pull request:
Opening as draft to see if this is something we want to do. If so, I'll finish it up (fixing tests, adding missing changelog entry, etc).

This PR includes the information currently only retrieved via `GetTransform` in the `Unit` message. This makes this data available everywhere we have units (unit stream, events, ...).

## Motivation

I am still working on an LSO, which was also the reason I've added `GetTransform` in the first place. I've now faced the challenge that I needed the information of `GetTransform` at the exact time of a land event. Sending a `GetTransform` after receiving the event yields data that is already newer than the event (units already moved since), which is why I need the information included in the event itself.

With this change, the information also ends up in the units stream, which would solve parts of #159.

## Drawback

More calculations and Lua calls done per unit export. Might effect the performance - though I am not sure if it is even a realistic effect to worry about?

rkusa commented on lua/DCS-gRPC/exporters/object.lua:
```diff
@@ -25,31 +25,61 @@ GRPC.exporters.position = function(pos)
+    headingDegrees = headingDegrees + 360
+  end
+
+  local velocity = object:getVelocity()
```
Most of the above calculations where previously done in Rust, moving them into Lua might effect the performance.

rurounijones replied:
Compared to the other stuff that runs in pure lua environments I am willing to take the risk. We can always run it against a complicated mission and see what the server reports.

rkusa commented on protos/dcs/common/v0/common.proto:
```diff
@@ -201,22 +201,17 @@ message Unit {
-  // The position of the unit
-  Position position = 6;
+  // Information about the unit in 3D space.
+  Transform transform = 6;
```
I am still unsure if `transform` is a good name, but still don't have a better idea.

rurounijones replied:
Works for me.

rkusa replied:
@omwdunkley called it _spatial_ [here](https://github.com/DCS-gRPC/rust-server/issues/159#issuecomment-1170409986), so maybe also a good candidate for a name? As a non-native English speaker I don't know which fits better. I'll keep `transform` unless someone else jumps in promoting a different name.

rurounijones replied:
I think transform works fine.

rkusa commented on protos/dcs/common/v0/common.proto:
```diff
@@ -201,22 +201,17 @@ message Unit {
-  double heading = 11;
   // The group category.
-  GroupCategory category = 12;
+  GroupCategory category = 10;
```
This is a breaking change that might effect some users of DCS-gRPC.

rurounijones replied:
Luckily we are in the 0.x.x 😁

rkusa commented on protos/dcs/common/v0/common.proto:
```diff
@@ -381,3 +375,44 @@ message Vector {
+  // The normalized direction the three line (right wing) is pointing to.
+  dcs.common.v0.Vector right = 2;
+  // The normalized up vector (orthogonal to forward and right).
+  dcs.common.v0.Vector up = 3;
```
Just a reminder, that `Vector` is still not a 1 to 1 mapping to DCS and rather a oppinionated modification of it (see https://github.com/DCS-gRPC/rust-server/blob/main/protos/dcs/common/v0/common.proto#L373-L383)

rurounijones replied:
I do not care personally. @bitbubbler - Opinions?

rkusa replied:
(I'd be fine with just using DCS' coordinate system and map them in my application rather than in DCS-gRPC)

bitbubbler replied:
Is there a good reason to not just expose the dcs system? If one want's to port scripts from lua to something that uses grpc why force them to adopt your opinionated system?
Unless it's solving a problem.

justin-lovell replied:
I reckon building a frequently calculated variables should be "opinionated." E.g. vectors, position, orientation.

However, limited to the discussion of the `Vector`, I agree with @bitbubbler as there are lots of DCS inspired Lua functions for calculating niche things, and I would be annoyed if I needed to change the vector calculations in the ported code in addition translating to another language (e.g. Lua to C#).

Let's expose those raw values (including the map's origin distance in meters) untouched, and where there is an opinion, it is extended on top of those raw values.

rkusa replied:
I'll update the code to use the raw values, thanks for the feedback everyeone 👍

rurounijones commented on lua/DCS-gRPC/exporters/object.lua:
```diff
@@ -25,31 +25,61 @@ GRPC.exporters.position = function(pos)
+      up = GRPC.exporters.vector(up),
+    },
+    velocity = GRPC.exporters.vector(velocity),
+    heading = headingDegrees,
+    speed = math.sqrt((velocity.x)^2+(velocity.z)^2),
```
Does this really need to be moved into the transform?

I am seeing "Heading" (which way nose is pointing) and "Speed" is the basic information that will probably work for most clients. OverlordBot, for example, will only care about heading and speed.

I see the information in the `Transform` section as being the "Advanced" stuff for those small number of people who actually care about it and want to do their own calculations while other clients can ignore its existance.

rkusa replied:
I personally need the `heading` as part of the `GetTransform` request, other than that, I'd be fine with moving heading and speed out of it. If we move them out of it, I'd probably still include them in the `GetTransform` response as follows:
```proto
Transform transform = 1;
double speed = 2;
double heading = 3;
```

rkusa replied:
Here is another idea. How about completely flattening the `transform` into the `Unit` message, so `unit.transform.velocity` becomes `unit.velocity` and so on. I'd then still tend to keep the `GetTransform` API call to get the subset of properties of the unit that describe it in 2D and 3D space (so basically everything currently in the `transform` message). Wdyt?

rkusa replied:
I'd like to redact my idea above. While working again on the PR, I've come to realize that I personally found it really useful that all those information are in one message/struct, as this allows them to be easily reused in different contexts (in the unit, in the dedicated method to get the transform, in the units stream).

rurounijones reviewed (commented):
Overall looks good to me.

justin-lovell commented:
I'm leaving a few thoughts on the API shapes which could help define a structure / context definition.

The semantics of gRPC appears to follow the theme of external observations. E.g., from a radar scope and movement context, heading is a velocity. 

In the definition of the "heading" in the other discussion where it is the nose position ("aiming point"), that is a concept which would only present from the internal/pilot instrumentation (e.g. HSI).

Therefore, the definition would have a vast difference depending on the context.

If the API is supposed to be opinionated from an external observer point of view, the API shape may be:

```
// physical position in space -- where a "dot" exists
position = { lat, lng, altitude }

// third party observation from a radar scope context (e.g. Overlord)
velocity = { heading, speed, ... and maybe fpm? ... }

// the physical "own ship" details.
orientation = { heading, roll, pitch }

// mapping for DCS
mapping= { up, forward, right }
```

Since there has been a use of the word "opinionated," there is merit in extracting a contextual definition. The `mapping` is an escape hatch for raw values if someone needs to do funky maths beyond the frequently derived variables (velocity, orientation, position)

rkusa commented on protos/dcs/mission/v0/mission.proto:
```diff
@@ -538,11 +538,14 @@ message StreamUnitsResponse {
+    dcs.common.v0.Unit new = 1;
+
+    // The unit's state changed (position, orientation, ...).
+    dcs.common.v0.Transform changed = 2;
```
By grouping everything related to 2D/3D position/transform into one message, we could easily distinguish in the units stream between new and changed, and skip re-sending static unit information again for the latter. Just in case this is something we want to do - I don't have a strong oppinion here myself, so I'd fine with keeping it as is.

rurounijones replied:
I would much prefer we do not do this. Doing so forces clients to maintain state now which not all clients will want to do. I would prefer we send the complete set of unit information, including static information, each time. I think the performance impact of this would be negligable.

rurounijones commented on protos/dcs/common/v0/common.proto:
```diff
@@ -380,13 +375,52 @@ message MarkPanel {
+message Transform {
+  dcs.common.v0.Position position = 1;
+  // Distance between DCS' map origin to object in meters on west-east axis.
+  double u = 2;
```
While I know this was requested I don't actually think it is useful information. I have never seen anyone figure out how to convert the internal coord system to lat/long outside of the DCS lua environment since some of the maps are skewed.

I think it was @justin-lovell who requested this. @justin-lovell, What is the use-case?

rkusa replied:
I need them when exporting units into a [ACMI (TacView) format](https://www.tacview.net/documentation/acmi/en/). Since DCS is basically just a flat world, I had way better results when using `u`/`v` compared to just using lat/lon.

rkusa commented:
@justin-lovell Thank you for your input! Some remarks:
- I kinda like the suggestion of moving `heading` and `speed` into `velocity`, and `heading` into `orientation` (as it is technically `yaw`) - though I am not sure if having two `heading` values is to confusing and we should rather keep the one in `orientation` named `yaw`?
- I'd propably prefer to keep `up, forward, right` inside of `orientation`  as it gives it some more context of what the values are (vectors describing the orientation in 3D space).
- For consistency, it would then also make sense to move `u` and `v` into `position`.

This would bring the structure to:

```lua
{
  position = { lat, lng, altitude, u, v },
  orientation = { yaw (or heading), roll, pitch, forward = { x, y, z }, right = { x, y, z }, up = { x, y, z } },
  velocity = { heading, speed, direction = { x, y, z } }, -- not sure if `direction` is a good name here for the vector that describes the velocity
}
```

What I still find a bit confusin is the mix of properties defining the object in 3D space, and the properties defining the object as if the world was flat. Values for the latter are: `position.{u,v}`, `velocity.heading`, `velocity.speed`. 🤔

rkusa commented:
I've made the following changes:

- Grouped the _transform` into:
  ```lua
  {
    position = { lat, lng, altitude, u, v },
    orientation = { heading, yaw, roll, pitch, forward = { x, y, z }, right = { x, y, z }, up = { x, y, z } },
    velocity = { heading, speed, velocity = { x, y, z } }
  }
  ```
- Added `position`/`orientation`/`velocity` directly to `Unit`, instead of grouping it into `transform` (solves my ongoing unhappiness with the term _transform_)
- Moved the math back into the Rust part of the code

I am now happy with the changes.

rkusa commented on protos/dcs/common/v0/common.proto:
```diff
@@ -171,6 +171,10 @@ message Position {
+  // Distance between DCS' map origin to object in meters on west-east axis.
+  double u = 4;
+  // Distance between DCS' map origin to object in meters on north-south axis.
+  double v = 5;
```
This adds `u`/`v` to all positions. I'd be up for changing the implementation to only include it for the position of units (might also be useful in the future for projectiles in case anyone wants to export them to TacView).

rurounijones replied:
👍

rkusa commented on stubs/src/common.rs:
```diff
@@ -0,0 +1,125 @@
+        raw_transform: Option<RawTransform>,
+    }
+
+    impl From<UnitIntermediate> for Unit {
```
This code allows to move the math back to Rust and automitcally apply it to all `Unit`s, regardless of where they are exported.

rurounijones replied:
Sounds like a good decision.

rkusa commented on stubs/src/lib.rs:
```diff
@@ -2,88 +2,25 @@
-        tonic::include_proto!("dcs.world.v0");
-    }
-}
+pub mod world;
```
I've moved all those `mod`s into a separate file. This improved the incremental compilation a lot when making changes to one of them.

rurounijones replied:
Very nice

rurounijones reviewed (approved):
On mobile so checking all the details is beyond at the moment but ddoximentatoon and whatnot are understandable so I think it is good from my end. Not sure if you want someone to have a closer look at the code.

rkusa commented:
Rebased due to merge conflicts. I'll be going ahead with merging this. I've tested the code in my LSO bot. I am happy to create follow-ups, if anything comes up (code-wise, or who the proto messages are structured).
